File: C++/Best Time to Buy and Sell Stock III.cpp
#include <vector>

int maxProfit(vector<int> &prices) {
    int len=prices.size();
    if(len<=1) return 0;
    int *dp1=new int[len];
    dp1[0]=0;
    int maxv=prices[0],minv=prices[0],maxp=0,maxnum=0;
    for(int i=0;i<len;i++){
        if(prices[i]>maxv){
            maxp=max(maxp,prices[i]-minv);
        }
        dp1[i]=maxp;
        if(prices[i]<minv){
            minv=prices[i];
            maxv=minv;
        }
        maxnum=max(maxnum,maxp);
    }
    int *dp2=new int[len];
    dp2[len-1]=0;
    maxv=prices[len-1],minv=prices[len-1],maxp=0;
    for(int i=len-1;i>=0;i--){
        if(prices[i]<minv){
            maxp=max(maxp,maxv-prices[i]);
        }
        dp2[i]=maxp;
        if(prices[i]>maxv){
            maxv=prices[i];
            minv=maxv;
        }
    }
    for(int i=0;i<len;i++){
        int cnt=dp1[i]+dp2[i];
        maxnum=max(maxnum,cnt);
    }
    return maxnum;
}

File: Python/Best-Time-to-Buy-and-Sell-Stock-III.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        length=len(prices)
        if length==0: return 0
        f1=[0 for i in range(length)]
        f2=[0 for i in range(length)]

        minV=prices[0]; f1[0]=0
        for i in range(1,length):
            minV=min(minV, prices[i])
            f1[i]=max(f1[i-1],prices[i]-minV)

        maxV=prices[length-1]; f2[length-1]=0
        for i in range(length-2,-1,-1):
            maxV=max(maxV,prices[i])
            f2[i]=max(f2[i+1],maxV-prices[i])

        res=0
        for i in range(length):
            if f1[i]+f2[i]>res: res=f1[i]+f2[i]
        return res